File: src/market_snapshot.py
"""Fetch a comprehensive market snapshot from Upstox REST APIs into data/.

Run AFTER a valid .access_token exists (auth.py or get_token.py):

    python src/market_snapshot.py

Pulls (best-effort - each section is independent, errors are captured not fatal):
  - Market status & exchange timings
  - Intraday (5m) + recent daily candles for the configured underlyings
  - Nearest-expiry option chain (OI / IV / greeks) for F&O underlyings
  - News headlines

Saves everything to data/snapshot_YYYYMMDD_HHMMSS.json for the analysis agents
(news-scanner, fno-strategist, market-analyst) to read.

NOTE: Historical candles work any time; option chain / live quotes are most
useful during or just after market hours.
"""

import os
import sys
import json
import datetime

import upstox_client
from upstox_client.rest import ApiException

# ============================================================
# CONFIG - edit underlyings / intervals here
# ============================================================
PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
TOKEN_FILE = os.path.join(PROJECT_ROOT, ".access_token")
DATA_DIR = os.path.join(PROJECT_ROOT, "data")

# Underlyings to analyse (index F&O by default; add equities as needed).
UNDERLYINGS = {
    "NIFTY": "NSE_INDEX|Nifty 50",
    "BANKNIFTY": "NSE_INDEX|Nifty Bank",
    "SENSEX": "BSE_INDEX|SENSEX",
}

INTRADAY_UNIT = "minutes"
INTRADAY_INTERVAL = "5"
HIST_UNIT = "days"
HIST_INTERVAL = "1"
HIST_LOOKBACK_DAYS = 40
# ============================================================


def to_serializable(obj):
    if obj is None:
        return None
    if hasattr(obj, "to_dict"):
        try:
            return obj.to_dict()
        except Exception:
            pass
    return str(obj)


def nearest_expiry(expiries, today_str):
    """Nearest tradeable expiry as bare YYYY-MM-DD (API wants date only).

    Prefers the nearest expiry strictly AFTER today (the next live weekly/monthly);
    falls back to today's, then to the earliest available.
    """
    dates = sorted({str(e).split(" ")[0] for e in expiries if e})
    future = [d for d in dates if d > today_str]
    if future:
        return future[0]
    same_or_future = [d for d in dates if d >= today_str]
    return same_or_future[0] if same_or_future else (dates[0] if dates else None)


def load_token():
    if not os.path.exists(TOKEN_FILE):
        print(f"ERROR: {TOKEN_FILE} not found. Run auth.py or get_token.py first.")
        sys.exit(1)
    with open(TOKEN_FILE, "r", encoding="utf-8") as f:
        token = f.read().strip()
    if not token:
        print(f"ERROR: {TOKEN_FILE} is empty. Re-authenticate.")
        sys.exit(1)
    return token


def main():
    token = load_token()
    cfg = upstox_client.Configuration()
    cfg.access_token = token
    client = upstox_client.ApiClient(cfg)

    snap = {
        "_generated_at": datetime.datetime.now(datetime.timezone.utc).isoformat(),
        "underlyings": UNDERLYINGS,
        "sections": {},
        "errors": {},
    }

    def grab(name, fn):
        """Run a fetch, store result or error - never abort the whole snapshot."""
        try:
            snap["sections"][name] = to_serializable(fn())
            print(f"  [ok]  {name}")
        except ApiException as e:
            snap["errors"][name] = f"ApiException {e.status}: {e.body}"
            print(f"  [ERR] {name}: HTTP {e.status}")
        except Exception as e:  # noqa: BLE001
            snap["errors"][name] = f"{type(e).__name__}: {e}"
            print(f"  [ERR] {name}: {e}")

    print("== Market status ==")
    grab(
        "market_status_NSE",
        lambda: upstox_client.MarketHolidaysAndTimingsApi(client).get_market_status("NSE"),
    )

    today = datetime.date.today().isoformat()
    from_date = (datetime.date.today() - datetime.timedelta(days=HIST_LOOKBACK_DAYS)).isoformat()
    hist = upstox_client.HistoryV3Api(client)

    print("== Candles ==")
    for name, key in UNDERLYINGS.items():
        grab(
            f"intraday_{name}",
            lambda key=key: hist.get_intra_day_candle_data(key, INTRADAY_UNIT, INTRADAY_INTERVAL),
        )
        grab(
            f"historical_{name}",
            lambda key=key: hist.get_historical_candle_data1(
                key, HIST_UNIT, HIST_INTERVAL, today, from_date=from_date
            ),
        )

    print("== Option chains ==")
    opt = upstox_client.OptionsApi(client)
    for name, key in UNDERLYINGS.items():
        try:
            contracts = to_serializable(opt.get_option_contracts(key))
            snap["sections"][f"option_contracts_{name}"] = contracts
            # Best-effort: find the nearest expiry from the contracts payload.
            expiries = []
            data = contracts.get("data") if isinstance(contracts, dict) else None
            if isinstance(data, list):
                expiries = sorted(
                    {c.get("expiry") for c in data if isinstance(c, dict) and c.get("expiry")}
                )
            snap["sections"][f"expiries_{name}"] = expiries
            exp = nearest_expiry(expiries, today)
            snap["sections"][f"nearest_expiry_{name}"] = exp
            if exp:
                grab(
                    f"option_chain_{name}",
                    lambda key=key, exp=exp: opt.get_put_call_option_chain(key, exp),
                )
            else:
                snap["errors"][f"option_chain_{name}"] = "no expiry found in contracts"
            print(f"  [ok]  option_contracts_{name}")
        except Exception as e:  # noqa: BLE001
            snap["errors"][f"option_contracts_{name}"] = f"{type(e).__name__}: {e}"
            print(f"  [ERR] option_contracts_{name}: {e}")

    # News is sourced by the news-scanner agent (web search). The Upstox news API is
    # instrument-scoped (not a market-wide "category"), so it's intentionally omitted here.

    os.makedirs(DATA_DIR, exist_ok=True)
    ts = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
    path = os.path.join(DATA_DIR, f"snapshot_{ts}.json")
    with open(path, "w", encoding="utf-8") as f:
        json.dump(snap, f, indent=2, default=str)

    print(f"\nSaved snapshot -> {path}")
    print(f"Sections OK:    {list(snap['sections'].keys())}")
    if snap["errors"]:
        print(f"Sections w/err: {list(snap['errors'].keys())}")


if __name__ == "__main__":
    main()

File: src/main/java/com/zaheenmunshi/upstox/MarketSnapshot.java
package com.zaheenmunshi.upstox;

import com.upstox.ApiClient;
import com.upstox.ApiException;
import com.upstox.api.GetOptionContractResponse;
import com.upstox.api.InstrumentData;
import io.swagger.client.api.HistoryV3Api;
import io.swagger.client.api.MarketHolidaysAndTimingsApi;
import io.swagger.client.api.OptionsApi;
import org.threeten.bp.OffsetDateTime;

import java.nio.file.Files;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.TreeSet;

/**
 * Port of {@code market_snapshot.py}: fetch a comprehensive market snapshot from
 * the Upstox REST APIs into {@code data/snapshot_*.json}.
 *
 * Each section is independent and best-effort — a failure is captured in
 * {@code errors} rather than aborting the whole snapshot.
 */
public final class MarketSnapshot {

    // ---- CONFIG: edit underlyings / intervals here ------------------------
    private static final Map<String, String> UNDERLYINGS = new LinkedHashMap<>();
    static {
        UNDERLYINGS.put("NIFTY", "NSE_INDEX|Nifty 50");
        UNDERLYINGS.put("BANKNIFTY", "NSE_INDEX|Nifty Bank");
        UNDERLYINGS.put("SENSEX", "BSE_INDEX|SENSEX");
    }
    private static final String INTRADAY_UNIT = "minutes";
    private static final int INTRADAY_INTERVAL = 5;
    private static final String HIST_UNIT = "days";
    private static final int HIST_INTERVAL = 1;
    private static final int HIST_LOOKBACK_DAYS = 40;
    // -----------------------------------------------------------------------

    /** A fetch that may throw the SDK's checked {@link ApiException}. */
    @FunctionalInterface
    private interface Fetch { Object get() throws Exception; }

    private final Map<String, Object> sections = new LinkedHashMap<>();
    private final Map<String, Object> errors = new LinkedHashMap<>();

    private MarketSnapshot() {}

    public static void main(String[] args) throws Exception {
        new MarketSnapshot().run();
    }

    void run() throws Exception {
        Config cfg = Config.load();
        ApiClient client = UpstoxClients.authenticated(cfg.loadToken());

        String generatedAt = Instant.now().toString();

        System.out.println("== Market status ==");
        MarketHolidaysAndTimingsApi timings = new MarketHolidaysAndTimingsApi(client);
        grab("market_status_NSE", () -> timings.getMarketStatus("NSE"));

        String today = LocalDate.now().toString();
        String fromDate = LocalDate.now().minusDays(HIST_LOOKBACK_DAYS).toString();
        HistoryV3Api hist = new HistoryV3Api(client);

        System.out.println("== Candles ==");
        for (Map.Entry<String, String> e : UNDERLYINGS.entrySet()) {
            String name = e.getKey(), key = e.getValue();
            grab("intraday_" + name, () -> hist.getIntraDayCandleData(key, INTRADAY_UNIT, INTRADAY_INTERVAL));
            grab("historical_" + name, () -> hist.getHistoricalCandleData1(key, HIST_UNIT, HIST_INTERVAL, today, fromDate));
        }

        System.out.println("== Option chains ==");
        OptionsApi opt = new OptionsApi(client);
        for (Map.Entry<String, String> e : UNDERLYINGS.entrySet()) {
            String name = e.getKey(), key = e.getValue();
            try {
                GetOptionContractResponse contracts = opt.getOptionContracts(key, null);
                sections.put("option_contracts_" + name, toSerializable(contracts));

                List<String> expiries = new ArrayList<>(new TreeSet<>(extractExpiries(contracts)));
                sections.put("expiries_" + name, expiries);
                String exp = nearestExpiry(expiries, today);
                sections.put("nearest_expiry_" + name, exp);
                if (exp != null) {
                    grab("option_chain_" + name, () -> opt.getPutCallOptionChain(key, exp));
                } else {
                    errors.put("option_chain_" + name, "no expiry found in contracts");
                }
                System.out.println("  [ok]  option_contracts_" + name);
            } catch (Exception ex) {
                errors.put("option_contracts_" + name, ex.getClass().getSimpleName() + ": " + ex.getMessage());
                System.out.println("  [ERR] option_contracts_" + name + ": " + ex.getMessage());
            }
        }

        // News is sourced by the news-scanner agent (web search); the Upstox news
        // API is instrument-scoped, so it is intentionally omitted here.

        // Compact, Java-computed summary so agents read this instead of re-deriving
        // indicators / option-chain stats from the raw sections (token + stability win).
        // Emitted near the top of the file; raw sections are left untouched below it.
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("_generated_at", generatedAt);
        out.put("underlyings", UNDERLYINGS);
        out.put("digest", SnapshotDigest.build(UNDERLYINGS, sections));
        out.put("sections", sections);
        out.put("errors", errors);

        Files.createDirectories(cfg.dataDir());
        String ts = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"));
        var path = cfg.dataDir().resolve("snapshot_" + ts + ".json");
        Json.MAPPER.writeValue(path.toFile(), out);

        System.out.println("\nSaved snapshot -> " + path);
        System.out.println("Sections OK:    " + sections.keySet());
        if (!errors.isEmpty()) {
            System.out.println("Sections w/err: " + errors.keySet());
        }
    }

    /** Run a fetch, store its (serialized) result or capture the error — never abort. */
    private void grab(String name, Fetch fn) {
        try {
            sections.put(name, toSerializable(fn.get()));
            System.out.println("  [ok]  " + name);
        } catch (ApiException api) {
            errors.put(name, "ApiException " + api.getCode() + ": " + api.getResponseBody());
            System.out.println("  [ERR] " + name + ": HTTP " + api.getCode());
        } catch (Exception ex) {
            errors.put(name, ex.getClass().getSimpleName() + ": " + ex.getMessage());
            System.out.println("  [ERR] " + name + ": " + ex.getMessage());
        }
    }

    /** Convert an SDK response into a plain JSON-friendly tree; fall back to toString. */
    private static Object toSerializable(Object obj) {
        if (obj == null) return null;
        try {
            return Json.MAPPER.convertValue(obj, Object.class);
        } catch (Exception e) {
            return String.valueOf(obj);
        }
    }

    private static List<String> extractExpiries(GetOptionContractResponse contracts) {
        List<String> out = new ArrayList<>();
        List<InstrumentData> data = contracts.getData();
        if (data == null) return out;
        for (InstrumentData inst : data) {
            OffsetDateTime exp = inst.getExpiry();
            if (exp != null) out.add(exp.toLocalDate().toString());
        }
        return out;
    }

    /**
     * Nearest tradeable expiry as bare YYYY-MM-DD. Prefers the nearest expiry
     * strictly after today; falls back to today's, then the earliest available.
     */
    static String nearestExpiry(List<String> sortedDates, String today) {
        List<String> dates = new ArrayList<>(new LinkedHashSet<>(sortedDates));
        for (String d : dates) if (d.compareTo(today) > 0) return d;   // first future
        for (String d : dates) if (d.compareTo(today) >= 0) return d;  // first same-or-future
        return dates.isEmpty() ? null : dates.get(0);                  // earliest
    }
}
